compiler/wasm/src/compile_new.rs

```diff
@@ -0,0 +1,258 @@
+    //
+    // This essentially says that CrateId1 depends on CrateId2 and the dependency is named `crate_name`
+    pub fn add_dependency_edge(
+        &mut self,
+        crate_name: String,
+        from: CrateIDWrapper,
+        to: CrateIDWrapper,
+    ) {
+        let parsed_crate_name: CrateName = crate_name
+            .parse()
+            .unwrap_or_else(|_| panic!("Failed to parse crate name {}", crate_name));
+        add_dep(&mut self.context, from.0, to.0, parsed_crate_name);
+    }
+
+    pub fn compile_program(mut self) -> Result<JsCompileResult, JsCompileError> {
+        let compile_options = CompileOptions::default();
+        // For now we default to plonk width = 3, though we can add it as a parameter
+        let np_language = acvm::Language::PLONKCSat { width: 3 };
+        #[allow(deprecated)]
+        let is_opcode_supported = acvm::pwg::default_is_opcode_supported(np_language);
+
+        let root_crate_id = *self.context.root_crate_id();
+
+        let compiled_program =
+            compile_main(&mut self.context, root_crate_id, &compile_options, None, true)
+                .map_err(|errs| {
+                    CompileError::with_file_diagnostics(
+                        "Failed to compile program",
+                        errs,
+                        &self.context.file_manager,
+                    )
+                })?
+                .0;
+
+        let optimized_program =
+            nargo::ops::optimize_program(compiled_program, np_language, &is_opcode_supported)
+                .expect("Program optimization failed");
+
+        let compile_output = preprocess_program(optimized_program);
+        Ok(JsCompileResult::new(compile_output))
+    }
+
+    pub fn compile_contract(mut self) -> Result<JsCompileResult, JsCompileError> {
+        let compile_options = CompileOptions::default();
+        // For now we default to plonk width = 3, though we can add it as a parameter
+        let np_language = acvm::Language::PLONKCSat { width: 3 };
+        #[allow(deprecated)]
+        let is_opcode_supported = acvm::pwg::default_is_opcode_supported(np_language);
+
+        let root_crate_id = *self.context.root_crate_id();
+
+        let compiled_contract =
+            compile_contract(&mut self.context, root_crate_id, &compile_options)
+                .map_err(|errs| {
+                    CompileError::with_file_diagnostics(
+                        "Failed to compile contract",
+                        errs,
+                        &self.context.file_manager,
+                    )
+                })?
+                .0;
+
+        let optimized_contract =
+            nargo::ops::optimize_contract(compiled_contract, np_language, &is_opcode_supported)
+                .expect("Contract optimization failed");
+
+        let compile_output = preprocess_contract(optimized_contract);
+        Ok(JsCompileResult::new(compile_output))
```

Comment:
Since we have separate methods for compiling contracts and compiling programs, this CompileResult is not as needed. Though its likely fine I think

Reply:
I think it makes sense to expose `compile_program` and `compile_contract` instead of the generic `compile` function that takes a boolean.